File: SliverpermissionX/src/main/java/com/gxinyu/sliverpermission/RequestFragment.java
package com.gxinyu.sliverpermission;

import android.app.Activity;
import android.app.Fragment;
import android.content.Intent;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import java.util.HashSet;
import java.util.Set;

public class RequestFragment extends Fragment {

    private Request mRequestManager;
    private RequestFragment rootRequestManagerFragment;
    private final Set<RequestFragment> childRequestManagerFragments = new HashSet<>();

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setRetainInstance(true);
    }

    @SuppressWarnings("deprecation")
    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        try {
            registerFragmentWithRoot(activity);
        } catch (IllegalStateException e) {

        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        unregisterFragmentWithRoot();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        unregisterFragmentWithRoot();
    }

    public Request getRequestManager() {
        return mRequestManager;
    }

    public void setRequestManager(Request requestManager) {
        mRequestManager = requestManager;
    }

    @SuppressWarnings("deprecation")
    private void addChildRequestManagerFragment(RequestFragment child) {
        childRequestManagerFragments.add(child);
    }

    @SuppressWarnings("deprecation")
    private void removeChildRequestManagerFragment(RequestFragment child) {
        childRequestManagerFragments.remove(child);
    }

    public void setParentFragmentHint(Fragment parentFragmentHint) {
        if (parentFragmentHint != null && parentFragmentHint.getActivity() != null) {
            registerFragmentWithRoot(parentFragmentHint.getActivity());
        }
    }

    @SuppressWarnings("deprecation")
    private void registerFragmentWithRoot(@NonNull Activity activity) {
        unregisterFragmentWithRoot();
        rootRequestManagerFragment = SliverPermission.get().getRequestRetriever().getRequestFragment(activity);
        if (!equals(rootRequestManagerFragment)) {
            rootRequestManagerFragment.addChildRequestManagerFragment(this);
        }
    }

    private void unregisterFragmentWithRoot() {
        if (rootRequestManagerFragment != null) {
            rootRequestManagerFragment.removeChildRequestManagerFragment(this);
            rootRequestManagerFragment = null;
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        mRequestManager.onActivityResult(requestCode, resultCode, data);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        mRequestManager.onRequestPermissionsResult(requestCode,permissions, grantResults);
    }


}

File: sliverpermissionkt/src/main/java/com/gxinyu/sliverpermission/RequestFragment.kt
package com.gxinyu.sliverpermission

import android.app.Activity
import android.app.Fragment
import android.content.Intent
import android.os.Build
import android.os.Bundle
import androidx.annotation.RequiresApi
import java.util.*

class RequestFragment : Fragment() {

    var requestManager: Request? = null
    private var rootRequestManagerFragment: RequestFragment? = null
    private val childRequestManagerFragments: MutableSet<RequestFragment> = HashSet()
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        retainInstance = true
    }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        try {
            registerFragmentWithRoot(activity)
        } catch (e: IllegalStateException) {
        }
    }

    override fun onDetach() {
        super.onDetach()
        unregisterFragmentWithRoot()
    }

    override fun onDestroy() {
        super.onDestroy()
        unregisterFragmentWithRoot()
    }

    private fun addChildRequestManagerFragment(child: RequestFragment) {
        childRequestManagerFragments.add(child)
    }

    private fun removeChildRequestManagerFragment(child: RequestFragment) {
        childRequestManagerFragments.remove(child)
    }

    fun setParentFragmentHint(parentFragmentHint: Fragment?) {
        if (parentFragmentHint != null && parentFragmentHint.activity != null) {
            registerFragmentWithRoot(parentFragmentHint.activity)
        }
    }

    private fun registerFragmentWithRoot(activity: Activity) {
        unregisterFragmentWithRoot()
        rootRequestManagerFragment = SliverPermission.INSTANCE.getRequestRetriever().getRequestFragment(activity)
        if (!equals(rootRequestManagerFragment)) {
            rootRequestManagerFragment!!.addChildRequestManagerFragment(this)
        }
    }

    private fun unregisterFragmentWithRoot() {
        if (rootRequestManagerFragment != null) {
            rootRequestManagerFragment!!.removeChildRequestManagerFragment(this)
            rootRequestManagerFragment = null
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent) {
        super.onActivityResult(requestCode, resultCode, data)
        requestManager!!.onActivityResult(requestCode, resultCode, data)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        requestManager!!.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }
}